draft-trammell-taps-interface.md

```diff
@@ -1073,117 +859,117 @@ decrease.
+to the Protocol Stack. It consists of a Framer object with a single Action,
+Frame. Since the Framer depends on the protocol used at the application layer,
+it is bound to the Preconnection during the pre-establishment phase:
 
-Connection.FrameWith(Framer)
+Preconnection.FrameWith(Framer)
 
-OctetArray := Framer.Frame(Content)
+OctetArray := Framer.Frame(Message)
 
 Sender-side framing is a convenience feature of the interface, for parity with
 receiver-side framing (see {{receive-framing}}).
 
 # Receiving Data {#receiving}
 
-Once a Connection is established, Content may be received on it. The interface
-notifies the application that content has been received via the Received
-event:
+Once a Connection is established, Messages may be received on it. The application can indicate that it is ready to receive Messages by calling Receive() on the connection.
+
+Connection.Receive(ReceiveHandler)
 
-Connection -> Received&lt;Content>
+Receive takes a single object, a ReceiveHandler which can handle the Received
+event and the ReceiveError error. Each call to Receive will result in at most
+one Received event being sent to the handler, though implementations may provide
+convenience functions to indicate readiness to receive a larger but finite
+number of Messages with a single call. This allows an application to provide
+backpressure to the transport stack when it is temporarily not ready to receive
+messages.
 
-As with sending, the type of the Content to be passed is dependent on the
+Connection -> Received&lt;Message>
+
+As with sending, the type of the Message to be passed is dependent on the
 implementation, and on the constraints on the Protocol Stacks implied by the
-Connection's transport parameters. The Content may also contain metadata from
+Connection's transport parameters. The Message may also contain metadata from
 protocols in the Protocol Stack; which metadata is available is Protocol Stack
 dependent. In particular, when this information is available, the value of the
 Explicit Congestion Notification (ECN) field is contained in such metadata.
 This information can be used for logging and debugging purposes, and for
 building applications which need access to information about the transport
 internals for their own operation.
 
-The Content object must provide some method to retrieve an octet array
+The Message object must provide some method to retrieve an octet array
 containing application data, corresponding to a single message within the
 underlying Protocol Stack's framing.  See {{receive-framing}} for handling
 framing in situations where the Protocol Stack provides octet-stream transport
 only.
 
+Note that as with sending, the Message object may represent a partial message,
+larger than the maximum message size than can be received atomically. In this
+case, the Message object passed contains some reference to the full Message it
+belongs to, as well as the byte range of the data content within the partial
+Message.
+
```

Comment:
What the receiver gets with this call may or may not coincide with what the sender has handed over as a message. The application must be able to figure out where a message begins and ends. No problem to do this generically, via framers and de-framers, but simply receiving a "message" doesn't tell the receiver anything about the sender's message size, and hence this paragraph makes no sense to me.

Reply:
Partial messages are different than deframed messages. However, the two play nicely together: if you have a non-framing Protocol Stack with vanilla TCP on top, and you have a deframer that knows how to split the stream into application layer messages, you might still end up in a situation where the framer says "hey, I got a length-prefixed thing that says it's 24MB" and the system only supports buffering messages 512kB at a time. So the deframer would chop the stream up into buffer-sized Partials and pass them up.

(Indeed, in the degenerate case where you're running on vanilla TCP and you don't pass a deframer down, the system default deframer would treat the entire stream as a single giant message, and pass partials down according to its own buffering and scheduling requirements)...

Not really clear what the problem is. Suggest we defer to London.

Reply:
I'm fine with deferring the discussion to London, but until then I consider this paragraph as wrong and don't think it should be added. We're trying to kick out stuff for which we don't have consensus - so: first convince me, then add it, not vice versa.

Reply:
I'd really like to see this left in, and I thought it had been discussed and agreed in Cambridge? Either way, let's discuss more for sure, but I think is necessary to have a document that actually works for these protocol use cases. If we decide some other way, we can change it, but I think we need this text in here for London.

@mwelzl one point to re-iterate is that the notion of Message does not need to match between the two sides—it's just a matter of how a local side parses bytes on a connection into chunks.

Reply:
so it's clear that we need partial message support on both send and receive, simply because there are buffers in the world. we handwaved on this wildly in previous revisions of this document (and allowed ourselves to do so because "content"). this corrects that.

I... really don't understand your objection here, Michael ("this paragraph is bad" is not extremely helpful). Can you help me out a bit? Is the problem the interplay between deframers and partial messages? Is the text just bad?

Reply:
Okay. As it's written it doesn't make sense to me - I absolutely agree with having stuff that actually works, but so does NEAT, and NEAT doesn't need this...  maybe I'm missing out on one of your use cases.

Brian's "indeed, ..." bracket above describes exactly what's on my mind: in this case, why would you need to consider this a partial message, and give a "reference to the full message it belongs to"?  This seems unnecessary to me.  Can we agree on phrasing this as optional? Phrasing it as factual (as it is now) indicates that there is no other way, which would IMO mean "no vanilla TCP". I would be fine with "In this case, the Message object passed may contain some reference to the full Message it belongs to, as well as the byte range of the data content within the partial Message."

Reply:
735da31582ec00bbebfd9cd4fa5d04f70a6bad61 contains an attempt to make this more explicit (and to note that vanilla TCP is of course supported). If making "full message reference" optional is better, happy to do that.